text/0011-validate-requests.md

```diff
@@ -0,0 +1,72 @@
+- Implementation PR(s):
+
+## Timeline
+
+- Start Date: 2021-06-23
+- Date Entered FCP:
+- Date Accepted:
+- Date Landed:
+
+## Summary
+
+Given a schema, _validation_ checks that policies conform to the schema, and _schema-based parsing_ checks that entities conform to the schema.
+This RFC proposes to add support for checking that a query conforms to a schema.
+
+## Basic example
+
+Say the schema specifies that an action `readFile` should only be used with principals of type `User` and resources of type `File`.
+This RFC proposes to change the signature of the `Request` constructor to add an optional schema argument that can be used to check  argument types when constructing a `Request`.
+
+Examples (w/ invalid syntax for EUIDs):
+
+```rust
+Request::new(Some(principal), Some(action), Some(resource), context, None); // no schema; works like before
+Request::new(Some(User::"alice"), Some(Action::"readFile"), Some(File::"secret_file.txt"), context, Some(schema)); // ok
+Request::new(Some(User::"alice"), Some(Action::"readFile"), Some(User::"bob"), context, Some(schema)); // error, invalid resource
+```
+
+## Motivation
+
+(Example migrated from #5)
+
+Say that a user wants to allow any `User` to view files in the folder `Folder::"public"`.
+Currently, to achieve the desired behavior, the user should define a schema with the action entity `readFile` that can apply to principals of type `User` and resources of type `File`, and then write the following policy.
+
+```
+permit(principal, action == Action::"readFile", resource in Folder::"public");
+```
+
+Assuming that they pass in inputs where the principal is of type `User` and the resource is of type `File`, this will work as expected.
+However, it could also return "allow" for other principal types like `Folder` or resource types like `User`.
+This is a bug on the user's end: they shouldn’t make a query with the action `readFile` when the principal is not a `User` and the resource is not a `File`.
+They are breaking the contract they stated in their schema.
+Nevertheless, it might be confusing that such a query does not result in a "deny".
+
+This RFC proposes to add a check, when constructing the `Request` type, that the query is valid according to the schema.
+
+## Detailed design
+
+Validation and schema-based parsing are both "opt-in" in the sense that they are not required, so we'll the same for query validation.
+Validation is done by a separate API (`Validator::validate`), while schema-based parsing is done by passing a (optional) schema argument into various parsing APIs, e.g., `Entities::from_json_str` and `Context::from_json_str`.
+If the schema argument is `None`, then parsing does no validation.
+
+This RFC proposes to modify the constructor for the `Request` type to add an optional schema type, similar to schema-based parsing.
+
+## Drawbacks
+
+- This is a breaking change that will affect any users that are manually constructing `Request`s (which is likely all users).
```

Comment:
Does it have to be? For example, we have a `new_with_unknowns` constructor (https://github.com/cedar-policy/cedar/blob/f233431efb82a1323e35cba2023b34613b8c99e5/cedar-policy-core/src/ast/request.rs#L91); could we have a `new_with_validation` constructor? 

For that matter -- should this RFC apply to partial evaluation-ready `Request`s too?

Reply:
To answer your second question: yes, I think we'll want validation supported for partial `Request`s too, once those are exposed.

Reply:
> The downside is that this will break existing users; the benefit is that this will force them to be aware of the new feature and start validating their requests.

Good characterization of the tradeoff. IMO, I'd rather go with two APIs because I don't want to break existing users, and because I expect that the non-validation path is the common one. People will largely use the validation path for debugging.

How would you feel about having two `new` methods, one with 4 parameters and one with 5? (I'm pretty sure Rust supports overloading.) The five-parameter one would call validation, and then invoke the four-parameter one.

Reply:
IIRC, Rust doesn't support function overloading.

Reply:
Rust does not support function overloading.  You can use traits to give the same function different definitions on different types, but the signature has to match with the exception of the type `T` that's implementing the trait.

Reply:
I somewhat in favor of making the breaking change.

Reply:
Same.  It seems like a worthwhile change to force people to adapt to when updating to Cedar 3.x (whenever 3.x happens).

Reply:
I'm imagining that the actual performance overhead of request validation would be negligible, noting that the issue of validating the `context` is orthogonal (users can already do so or not as they construct the `Context` object which is passed to `Request::new()`), so the only performance overhead would be checking the `principal`, `action`, and `resource` are of compatible types. (Right?)

Reply:
To clarify: Users will validate the context if they do schema-based parsing to construct. They aren't required to do that.

Reply:
Correct.  It's orthogonal to whether they validate the request (which would only happen if the supply the schema to `Request::new()`).  They could do either one alone, or both, or neither.